app/com/linkedin/drelephant/tuning/hbt/PigHbtParameterRecommender.java

```diff
@@ -0,0 +1,406 @@
+              latestAppliedParams.put(REDUCER_HEAP_HADOOP_CONF.getValue(), getHeapMemory(heuristicResultDetail.value));
+            }
+          }
+        }
+      }
+    } else {
+      throw new IllegalArgumentException("No App Results found while TuneIn was suggesting new parameters");
+    }
+  }
+
+  /**
+   * This method is responsible for suggesting the parameters to fix failed heuristics
+   * in the next execution or if there is no failed heuristics then try to optimize Resource usage
+   * @return Map with suggested parameter's name as key and its value
+   */
+  public Map<String, Double> suggestParameters() {
+    loadLatestAppliedParameters();
+    List<String> failedHeuristicNameList = getFailedHeuristics();
+    suggestToFixFailedHeuristics(failedHeuristicNameList);
+    if (failedHeuristicNameList.size() == 0) {
+      logger.info("No failed heuristics, so will try to optimize memory parameters");
+      suggestMemoryParam(MRJobTaskType.MAP);
+      suggestMemoryParam(MRJobTaskType.REDUCE);
+    }
+    addRequiredPreviousSuggestedParameters();
+    modifyMapperMemory();
+    return jobSuggestedParameters;
+  }
+  /**
+   * Method to suggest parameters accordingly for each failed heuristic
+   * @param failedHeuristicNameList list with names of heuristics which are failed
+   */
+  private void suggestToFixFailedHeuristics(List<String> failedHeuristicNameList) {
+    for (String failedHeuristicName : failedHeuristicNameList) {
+      switch (failedHeuristicName) {
+        case MAPPER_MEMORY :
+          suggestMemoryParam(MRJobTaskType.MAP);
+          break;
+        case REDUCER_MEMORY :
+          suggestMemoryParam(MRJobTaskType.REDUCE);
+          break;
+        case MAPPER_SPILL :
+          suggestParametersForMemorySpill();
+          break;
+        case MAPPER_TIME:
+        case MAPPER_SPEED:
+          suggestSplitSize();
+      }
+    }
+  }
+
+  /**
+   * Method to add previous applied parameters values for Parameters
+   * which are not suggested this time
+   */
+  private void addRequiredPreviousSuggestedParameters() {
+    for (String key : latestAppliedParams.keySet()) {
+      if (!jobSuggestedParameters.containsKey(key)) {
+        jobSuggestedParameters.put(key, latestAppliedParams.get(key));
+      }
+    }
+  }
+
+  /**
+   * Method to suggest Memory for container for provided @taskType
+   * @param taskType Task type which can be either Map or Reduce
+   */
+  @VisibleForTesting
+  void suggestMemoryParam(MRJobTaskType taskType) {
+    String memoryHeuristicName;
+    if (taskType.equals(MRJobTaskType.MAP)) {
+      memoryHeuristicName = MAPPER_MEMORY_HEURISTICS_CONF.getValue();
+    } else {
+      memoryHeuristicName = REDUCER_MEMORY_HEURISTICS_CONF.getValue();
+    }
+    double maxUsedPhysicalMapperMemoryByJob =
+        getHeuristicDetailsMaxValueForJob(memoryHeuristicName, MAX_PHYSICAL_MEMORY.getValue());
```

Comment:
Why to call this method repeatedly . Cann't we get all desired max in one go , please explain

Reply:
Notable suggestion.
Have made the changes for fetching max once only.